src/vmm/src/arch/x86_64/mod.rs

```diff
@@ -110,13 +125,146 @@ pub fn initrd_load_addr(
+    add_memmap_entry(&mut memmap, 0, layout::SYSTEM_MEM_START, MEMMAP_TYPE_RAM)?;
+    add_memmap_entry(
+        &mut memmap,
+        layout::SYSTEM_MEM_START,
+        layout::SYSTEM_MEM_SIZE,
+        E820_RESERVED,
+    )?;
+    let last_addr = guest_mem.last_addr();
+    if last_addr < end_32bit_gap_start {
+        add_memmap_entry(
+            &mut memmap,
+            himem_start.raw_value(),
+            last_addr.unchecked_offset_from(himem_start) + 1,
+            MEMMAP_TYPE_RAM,
+        )?;
+    } else {
+        add_memmap_entry(
+            &mut memmap,
+            himem_start.raw_value(),
+            end_32bit_gap_start.unchecked_offset_from(himem_start),
+            MEMMAP_TYPE_RAM,
+        )?;
+
+        if last_addr > first_addr_past_32bits {
+            add_memmap_entry(
+                &mut memmap,
+                first_addr_past_32bits.raw_value(),
+                last_addr.unchecked_offset_from(first_addr_past_32bits) + 1,
+                MEMMAP_TYPE_RAM,
+            )?;
+        }
+    }
+
+    // Construct the hvm_start_info structure and serialize it into
+    // boot_params.  This will be stored at PVH_INFO_START address, and %rbx
+    // will be initialized to contain PVH_INFO_START prior to starting the
+    // guest, as required by the PVH ABI.
+    let mut start_info = hvm_start_info {
+        magic: XEN_HVM_START_MAGIC_VALUE,
+        version: 1,
+        cmdline_paddr: cmdline_addr.raw_value(),
+        memmap_paddr: layout::MEMMAP_START,
+        memmap_entries: memmap.len() as u32,
+        nr_modules: modules.len() as u32,
+        ..Default::default()
+    };
+    if !modules.is_empty() {
+        start_info.modlist_paddr = layout::MODLIST_START;
+    }
+    let mut boot_params =
+        BootParams::new::<hvm_start_info>(&start_info, GuestAddress(layout::PVH_INFO_START));
+
+    // Copy the vector with the memmap table to the MEMMAP_START address
+    // which is already saved in the memmap_paddr field of hvm_start_info struct.
+    boot_params.set_sections::<hvm_memmap_table_entry>(&memmap, GuestAddress(layout::MEMMAP_START));
+
+    // Copy the vector with the modules list to the MODLIST_START address.
+    // Note that we only set the modlist_paddr address if there is a nonzero
+    // number of modules, but serializing an empty list is harmless.
+    boot_params.set_modules::<hvm_modlist_entry>(&modules, GuestAddress(layout::MODLIST_START));
+
+    // Write the hvm_start_info struct to guest memory.
+    PvhBootConfigurator::write_bootparams(&boot_params, guest_mem)
+        .map_err(|_| ConfigurationError::StartInfoSetup)
+}
+
+fn add_memmap_entry(
+    memmap: &mut Vec<hvm_memmap_table_entry>,
+    addr: u64,
+    size: u64,
+    mem_type: u32,
+) -> Result<(), ConfigurationError> {
+    // Add the table entry to the vector
+    memmap.push(hvm_memmap_table_entry {
+        addr,
+        size,
+        type_: mem_type,
+        reserved: 0,
+    });
+
+    Ok(())
+}
```

Comment:
just inline this

Reply:
I don't really see any strong argument for doing it either way. I kinda like this admittedly because calling there's less visual clutter. Inlining it would also mean doing the constructing the struct at all 8 call sites, which involves explicitly writing out all fields, because outside of this functions the local variable names don't match up with the struct names.I

It doesnt need to return `Result` though, I'll give you that.

Reply:
@bchalios heya, could you weigh in here to resolve our deadlock? :D

Reply:
Personally, I would just call `memap.push` directly and get rid of this function all together. Wrapping this in a function doesn't add any value whatsoever. There's nothing to abstract and it doesn't really remove that much of clutter :)

Reply:
Okay, done